File: src/main/java/editor/filter/leaf/options/single/ExpansionFilter.java
package editor.filter.leaf.options.single;

import java.util.Arrays;
import java.util.HashSet;

import com.google.gson.JsonElement;
import com.google.gson.JsonPrimitive;

import editor.database.attributes.CardAttribute;
import editor.database.attributes.Expansion;
import editor.database.card.Card;
import editor.filter.Filter;

/**
 * This class represents a filter that groups cards by expansion.
 *
 * @author Alec Roelke
 */
public class ExpansionFilter extends SingletonOptionsFilter<Expansion>
{
    /**
     * Create a new ExpansionFilter.
     */
    public ExpansionFilter()
    {
        super(CardAttribute.EXPANSION, Card::expansion);
    }

    @Override
    protected Expansion convertFromString(String str)
    {
        for (Expansion expansion : Expansion.expansions)
            if (str.equalsIgnoreCase(expansion.name()))
                return expansion;
        throw new IllegalArgumentException("Unknown expansion name \"" + str + "\"");
    }

    @Override
    public Filter copy()
    {
        ExpansionFilter filter = (ExpansionFilter)CardAttribute.createFilter(CardAttribute.EXPANSION);
        filter.contain = contain;
        filter.selected = new HashSet<>(selected);
        return filter;
    }

    @Override
    protected JsonElement convertToJson(Expansion item)
    {
        return new JsonPrimitive(item.name());
    }

    @Override
    protected Expansion convertFromJson(JsonElement item)
    {
        return Arrays.stream(Expansion.expansions).filter((e) -> e.name().equalsIgnoreCase(item.getAsString())).findAny().get();
    }
}

File: src/main/scala/editor/filter/leaf/options/single/ExpansionFilter.scala
package editor.filter.leaf.options.single

import editor.database.attributes.CardAttribute
import editor.database.attributes.Expansion
import editor.filter.FaceSearchOptions
import editor.util.Containment

/**
 * Filter that groups cards by expansion.
 * @author Alec Roelke
 */
final case class ExpansionFilter(contain: Containment = Containment.AnyOf, selected: Set[Expansion] = Set.empty) extends SingletonOptionsFilter[Expansion, ExpansionFilter] {
  override def faces = FaceSearchOptions.ANY
  override def attribute = CardAttribute.Expansion
  override val unified = true
  override def value = _.expansion
  override def copy(faces: FaceSearchOptions, contain: Containment, selected: Set[Expansion]) = copy(contain = contain, selected = selected)
}